opm/autodiff/WellStateFullyImplicitBlackoil.hpp

```diff
@@ -282,6 +293,196 @@ namespace Opm
+                    // improved during the solveWellEq process
+                    {
+                        // top segment is always the first one, and its pressure is the well bhp
+                        segpress_.push_back(bhp()[w]);
+                        const int top_segment = top_segment_loc_[w];
+                        const int start_perf = wells->well_connpos[w];
+                        for (int seg = 1; seg < nseg; ++seg) {
+                            if ( !segment_perforations[seg].empty() ) {
+                                const int first_perf = segment_perforations[seg][0];
+                                segpress_.push_back(perfPress()[start_perf + first_perf]);
+                            } else {
+                                // segpress_.push_back(bhp); // may not be a good decision
+                                // using the outlet segment pressure // it needs the ordering is correct
+                                const int outlet_seg = segment_set[seg].outletSegment();
+                                segpress_.push_back(segpress_[top_segment + segment_set.numberToLocation(outlet_seg)]);
+                            }
+                        }
+                    }
+                }
+            }
+            assert(int(segpress_.size()) == nseg_);
+            assert(int(segrates_.size()) == nseg_ * numPhases() );
+
+            if (!prev_well_state.wellMap().empty()) {
+                // copying MS well related
+                const auto& end = prev_well_state.wellMap().end();
+                const int np = numPhases();
+                for (int w = 0; w < nw; ++w) {
+                    const std::string name( wells->name[w] );
+                    const auto& it = prev_well_state.wellMap().find( name );
+
+                    if (it != end) { // the well is found in the prev_well_state
+                        // TODO: the well with same name can change a lot, like they might not have same number of segments
+                        // we need to handle that later.
+                        // for now, we just copy them.
+                        const int old_index_well = (*it).second[0];
+                        const int new_index_well = w;
+                        const int old_top_segment_location = prev_well_state.topSegmentLocation(old_index_well);
+                        const int new_top_segmnet_location = topSegmentLocation(new_index_well);
+                        int number_of_segment = 0;
+                        if (new_index_well == int(top_segment_loc_.size()) - 1) {
+                            number_of_segment = nseg_ - new_top_segmnet_location;
+                        } else {
+                            number_of_segment = topSegmentLocation(new_index_well + 1) - new_top_segmnet_location;
+                        }
+
+                        for (int i = 0; i < number_of_segment * np; ++i) {
```

Comment:
This is ok only if the number of segments is constant in time. Is it?

Reply:
I think so. As indicated in the comments slightly earlier. 
We have not implemented the case to handle dynamically changed segment structure yet. 

Things can be more complicated, like same number of segments, while quite different segment structure. Currently, for the standard well, we also have this problem, like number of perforations are the same, the well can contain different perforations.  A good way is probably to investigate the usage of the Event from the opm-parser, while have not looked into it. It was implemented there months ago.

Reply:
> We have not implemented the case to handle dynamically changed segment structure yet.

Ok then this is ok for now.